src/iop/basicadj.c

```diff
@@ -830,14 +821,13 @@ static inline int ilogbp1(double d)
   return q;
 }
 
-static inline double ldexpk(double x, int q)
+// calculate  x * 2^q
+static inline double ldexpk(double x, int32_t q)
 {
-  double u;
-  int m;
-  m = q >> 31;
+  int32_t m = q < 0 ? 1 : 0;
```

Comment:
1. there is nothing wrong with the original code
2. the new code is blatantly wrong https://godbolt.org/z/Pvf7ao

Reply:
1. I'm not sure how it's in C, but in C++ until C++-20 results of right-shift of negative values is implementation-defined, so with >>31 you may get non-negative number...
2. Indeed

Reply:
Quoting ISO/IEC 9899:1999 (E)

> The result of E1 >> E2 is E1 right-shifted E2 bit positions. If E1 has an unsigned type
or if E1 has a signed type and a nonnegative value, the value of the result is the integral
part of the quotient of E1 / 2^E2 . If E1 has a signed type and a negative value, the
resulting value is implementation-defined.

I am not aware of that this would have changed in more recent versions of the C programming language.

Reply:
Still, it has to be `int32_t m = q < 0 ? -1 : 0;`